### PR TITLE
Feature/sprite optimized

### PR DESCRIPTION
**Pull request by Mee-gu**
sprite optimized

### CODE REVIEW

File: cocos/renderer/backend/metal/ShaderModuleMTL.h [hunk 1]
```diff
@@ -79,17 +78,30 @@ class ShaderModuleMTL : public ShaderModule
...
     
     id<MTLFunction> _mtlFunction = nil;
     
     std::unordered_map<std::string, UniformInfo> _uniformInfos;
+    std::unordered_map<int, UniformInfo> _activeUniformInfos;
     std::unordered_map<std::string, AttributeBindInfo> _attributeInfo;
     
     int _maxLocation = -1;
+
+    std::vector<char> _uniformBuffer;
```
Comment (minggo) [inline cocos/renderer/backend/metal/ShaderModuleMTL.h]: It is better to alloc memory without using std::vector<>.

File: cocos/renderer/backend/Types.cpp [hunk 2]
```diff
@@ -30,11 +30,24 @@ CC_BACKEND_BEGIN
...
+}
+
+void UniformLocation::reset() 
+{ 
+    location[0] = -1;
+#ifdef CC_USE_METAL
+    location[1] = -1; 
+#else
+    location[1] = 0;
+#endif
```
Comment (minggo) [inline cocos/renderer/backend/Types.cpp]: Can just set to -1 for metal and opengl?

File: cocos/2d/CCSprite.cpp [hunk 3]
```diff
@@ -379,6 +382,17 @@ void Sprite::updateShaders(const char* vert, const char* frag)
...
 
+void Sprite::setProgramState(backend::ProgramType type)
+{
+    if(_programState != nullptr &&
+       _programState->getProgram()->getProgramType() == type)
+        return;
+    
+    auto programState = new (std::nothrow) backend::ProgramState(type);
+    setProgramState(programState);
+    CC_SAFE_RELEASE_NULL(programState);
```
Comment (minggo) [inline cocos/2d/CCSprite.cpp:393]: CC_SAFE_RELEASE_NULL ->  CC_SAFE_RELEASE, don't have to set to null here.

File: cocos/renderer/backend/ProgramCache.h [hunk 4]
```diff
@@ -56,6 +56,9 @@ class ProgramCache : public Ref
      */
     backend::Program* newProgram(const std::string& vertexShader, const std::string& fragmentShader);
     
+    /// get built-in program
+    backend::Program* newProgramByProgramType(ProgramType type);
```
Comment (minggo) [inline cocos/renderer/backend/ProgramCache.h]: newProgramByProgramType -> newBuiltinProgram ?

File: cocos/renderer/backend/ProgramState.cpp [hunk 5]
```diff
@@ -263,8 +216,6 @@ ProgramState::~ProgramState()
 {
     CC_SAFE_RELEASE_NULL(_program);
```
Comment (minggo) [inline cocos/renderer/backend/ProgramState.cpp]: CC_SAFE_RELEASE_NULL -> CC_SAFE_RELEASE

File: cocos/renderer/backend/ProgramState.cpp [hunk 6]
```diff
@@ -263,8 +216,6 @@ ProgramState::~ProgramState()
...
     
-    _vertexUniformInfos.clear();
-    _fragmentUniformInfos.clear();
     _vertexTextureInfos.clear();
     _fragmentTextureInfos.clear();
```
Comment (minggo) [inline cocos/renderer/backend/ProgramState.cpp]: Don't need clear operations.

File: cocos/renderer/backend/ProgramState.h [hunk 7]
```diff
@@ -307,6 +300,9 @@ class ProgramState : public Ref
      * Reset uniform informations when EGL context lost
      */
     void resetUniforms();
+
+    ///Initialize.
+    void init();
```
Comment (minggo) [inline cocos/renderer/backend/ProgramState.h:307]: Should be public?

File: cocos/renderer/backend/metal/ShaderModuleMTL.h (hunk 1 above)
Comment (minggo) [reply cocos/renderer/backend/metal/ShaderModuleMTL.h]: Program doesn't need to have uniform buffer, just need the buffer length.

File: cocos/renderer/backend/ProgramState.h (hunk 7 above)
Comment (Mee-gu) [reply cocos/renderer/backend/ProgramState.h:307]: Not need to be public.

File: cocos/renderer/backend/Types.cpp (hunk 2 above)
Comment (Mee-gu) [reply cocos/renderer/backend/Types.cpp]: yes, it can do that.

File: cocos/renderer/backend/ProgramCache.h (hunk 4 above)
Comment (Mee-gu) [reply cocos/renderer/backend/ProgramCache.h]: Got it

**Pull request merged**